enhancements/0002-20221026-experimental-rest-transport-layer.md

```diff
@@ -0,0 +1,156 @@
+- *Support proxy servers.* Enable client/server connections to go through proxy servers that may or may not alter the request, for example, to conceal the client's IP address from the server.
+- *Keep client identity optional.* Clients should be able to remain anonymous, client identity should be an optional component that users of the framework can opt for depending on the needs of their workload.
+- *Prevent clients from poisoning the model.*
+- *Limit server-side knowledge about individual clients.*
+
+### Non-Goals
+
+- *End-to-end latency.* The existing bidirectional stream between client and server offers stellar end-to-end latency (i.e., from scheduling a task for a client until that client starts executing the task locally). Most workloads do not require such low levels of latency.
+- *API stability.* The proposed REST API is not intended to be final.
+- *Client authentication.* There should be a forward-compatible path for adding client authentication, but client authentication does not need to be implemented in the first version.
+- *Compatibility with `start_server`.*
+- *Build a "RESTful" API.* The goal is to have a REST-inspired API, but not to build a fully idiomatic RESTful API.
+
+## Proposal
+
+Introduce a request/response REST transport layer. The REST transport layer would consist of a REST API server (integrated with the Flower server) and a REST client.
+
+## Design Details
+
+### Definition(s)
+
+Task: User-defined specification of what a client has to do, with the expectation that the client returns a result.
+
+### Protocol
+
+General idea:
+- Client always initiates conversation
+- Server responds in one of two ways:
+  - Reconnect at a later time.
+  - Execute a task and send the result.
+- Server has a (mild) form of control over the conversation by suggesting to the client when it should connect again.
+
+Example:
+- Client connects to the server: "I'm available"
+- Server doesn't have any tasks: "reconnect in 5min"
+- Client sleeps for 5min, then connects again
+- Server say: here's a TASK and a TOKEN, work on it and send me your RESULT
+- Client receives the TASK, executes it locally, and returns the RESULT
+- Server says thanks, please reconnect in 10min
+- [the process repeats]
+
+### Server-side REST API
+
+Situation: a `Task`/`TaskAssignment` scheduled for `N` anonymous clients
+
+**GET /tasks**
+Request body: `None`
+Response body: `GetTasksResponse` containing either `Task` + `token` or `Reconnect`
+
+Implementation:
+1. Is there any task scheduled? Yes, one for `N` anonymous clients
+2. Generate the first `token`, store it with the `Task`/`TaskAssignment` in the `State`
+3. Return `Task` + `token` to the client in the `GetTasksResponse`
+
+**POST /results**
+
+Request body: `SubmitResultRequest` containing `Result` + `token`
+Response body: `None`
+
+Implementation:
+1. Check if the `token` is known
+2. Save `Result` if true, else return error
```

Comment:
If the token has expired and new task is in an early "distribution" phase we could distribute the update task with a new token since we indirectly know the client is available? This might be a way to recover the straggler, but not sure how much of an optimization this actually is

Reply:
That sounds interesting. We might introduce this check, and if (the token has expired and there are some tasks for that client) then send the list of tasks right away.

Reply:
Interesting, we discussed something along these lines where the server would reply in a way that tells the client what to do next. The most generic version of this might be a design where we only have one method: the client tells the server "hey, I'm here" or "hey, I'm here and I have these result/token pairs for you" and the server replies "reconnect later" or "here are one or more tasks, please work on them".